include/generic_sort.h

```diff
@@ -0,0 +1,190 @@
+/*********************                                                        */
+/*! \file generic_sort.h
+** \verbatim
+** Top contributors (to current version):
+**   Yoni Zohar
+** This file is part of the smt-switch project.
+** Copyright (c) 2020 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief Class that represents a sort in generic solvers
+**        
+**
+**/
+
+#pragma once
+
+#include "exceptions.h"
+#include "smt_defs.h"
+#include "sort.h"
+#include <functional>
+namespace smt {
+
+/* Helper functions for creating generic sorts */
+Sort make_uninterpreted_generic_sort(std::string name, uint64_t arity);
+Sort make_uninterpreted_generic_sort(Sort sort_cons, const SortVec& sorts);
+Sort make_generic_sort(SortKind sk);
+Sort make_generic_sort(SortKind sk, uint64_t width);
+Sort make_generic_sort(SortKind sk, Sort sort1);
+Sort make_generic_sort(SortKind sk, Sort sort1, Sort sort2);
+Sort make_generic_sort(SortKind sk, Sort sort1, Sort sort2, Sort sort3);
+Sort make_generic_sort(SortKind sk, SortVec sorts);
+
+/* smtlib representation of sort kinds */
+std::string to_smtlib(SortKind);
+
+/** \class GenericSort
+ *  An abstract class for generic Sorts
+ */
+class GenericSort : public AbsSort
+{
+ public:
+  GenericSort(SortKind sk); 
+  virtual ~GenericSort();
+  SortKind get_sort_kind() const override;
+  bool compare(const Sort & s) const override;
+
+  /** The following functions are 
+   *   implemented by sub-classes 
+   */
+
+  uint64_t get_width() const override
+  {
+    throw NotImplementedException(
+        "get_width not implemented by GenericSort");
+  }
+
+  uint64_t get_arity() const override
```

Comment:
This should be `size_t` actually -- that should also fix the Travis build. It is strange that `get_width` returns `uint64_t` and `get_arity` returns `size_t`. You're welcome to fix that but it's not your responsibility, I can go through and pick one later.